icu4c/source/extra/uconv/uconv.cpp

```diff
@@ -20,6 +20,10 @@
  * See source code repository history for changes.
  */
 
+#ifndef UCHAR_TYPE
+#define UCHAR_TYPE char16_t
+#endif
```

Comment:
umachine.h says
```
    // The intltest Makefile #defines UCHAR_TYPE=char16_t,
    // so we only #define it to uint16_t if it is undefined so far.
```
Should we define this in more makefiles, so that we don't need these definitions in the source files?
Or do you think it's more robust to do this in the source files, so that it works out of the box in any build system?

Reply:
Both choices seem unfortunate. Moving the def to makefiles would require separate defs  for Windows and regular make. But having all the repeated defs in the various source files also feels wrong.

I have no strong opinion. Leave it as proposed, and hope that some year we can remove all support for UChar being anything but char16_t, everywhere?

Reply:
Having that snippet in all those source files feels really bad to me, but I originally fixed this in Apple ICU under time pressure, so I went with the easiest thing I could think of.

I'd forgotten Windows has all different makefiles, and having to change them scares me, since I don't have any way to test it.

What I was originally thinking of was to add something analogous to `U_COMMON_IMPLEMENTATION` for the unit-test files and then just checking for that in `umachine.h`.  Although I guess that still amounts to messing with the makefiles.  Let me see if I can get that to work...

>...and hope that some year we can remove all support for UChar being anything but char16_t, everywhere?

Well, we're here because I inadvertently committed the unchanged OSICU code (with `#if 1`), and it caused a whole mess of build failures here at Apple.  So we have numerous projects here that still depend on `UChar` being `uint16_t`.  I should probably start working on people to change that dependency, but I don't know how quickly I can get that to happen.